### PR TITLE
`protostar declare`

### PR DESCRIPTION
**Pull request by kasperski95**
- added `declare` command
- added `Deployer` class that is responsible for declaring/deploying contracts
- connected `Deployer` class to `deploy` and `declare` commands
- updated deployment docs

### CODE REVIEW

File: website/docs/tutorials/guides/deploying.md [hunk 1]
```diff
@@ -4,9 +4,40 @@ sidebar_label: Deploying (1 min)
...
+## Declaring a contract
+StarkNet separates contracts into classes and instances. To create a new contract, you need to:
+1. declare a class
+2. call the deploy syscall with the `class_hash` from the declare transaction as an argument
+   ```cairo
+   from starkware.starknet.common.syscalls import deploy
+   ```
+   
+To declare a contract from the Protostar CLI you need to build your project and use the `protostar declare` command.
+
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/guides/deploying.md:20]:
```suggestion
## Declaring a contract
StarkNet provides a distinction between contract class and instance. To create a new contract, you need to:
1. Declare a contract
2. Use the deploy syscall with the `class_hash` from the declare transaction as an argument
   ```cairo
   from starkware.starknet.common.syscalls import deploy
   ```
   
To declare a contract from the Protostar CLI you need to build your project and use the `protostar declare` command.

```

File: website/docs/tutorials/guides/testing.md [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ---
-sidebar_label: Testing (6 min)
+sidebar_label: Testing (10 min)
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/guides/testing.md:2]: I don't think anybody will be able to read this page in 10 min. Can we just remove this time hints?

File: protostar/commands/declare/declare_command.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    @property
+    def example(self) -> Optional[str]:
+        return None
+
+    @property
+    def arguments(self) -> List[Command.Argument]:
+        return [
+            Command.Argument(
+                name="contract",
+                description=("The path to the compiled contract."),
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/declare/declare_command.py:35]:
```suggestion
                description=("A path to a compiled contract."),
```

File: protostar/commands/declare/declare_command.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                type="path",
+                is_positional=True,
+                is_required=True,
+            ),
+            Command.Argument(
+                name="signature",
+                description=("The signature information for the declaration."),
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/declare/declare_command.py:42]:
```suggestion
                description=("A signature information for the declaration."),
```

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: protostar/commands/declare/declare_command.py (hunk 3 above)
Comment (mkaput) [reply protostar/commands/declare/declare_command.py:35]: @MaksymilianDemitraszek is the `A` article necessary here?

File: protostar/deployer/deployer.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+from protostar.deployer.gateway_response import (
+    SuccessfulDeclareResponse,
+    SuccessfulDeployResponse,
+)
+from protostar.deployer.network_config import NetworkConfig
+from protostar.deployer.starkware.starknet_cli import deploy
+from protostar.protostar_exception import ProtostarException
+
+
+class InvalidNetworkConfigurationException(BaseException):
```
Comment (mkaput) [inline protostar/deployer/deployer.py:23]:
```suggestion
class InvalidNetworkConfigurationException(Exception):
```

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/deployer/deployer.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        compiled_contract_path: Path,
+        gateway_url: str,
+        signature: Optional[List[str]] = None,
+        token: Optional[str] = None,
+    ):
+        """Protostar version of starknet_cli::declare"""
+
+        # The following parameters are hardcoded because Starknet CLI have asserts checking if they are equal to these
+        # values. Once Starknet removes these asserts, these parameters should be configurable by the user.
+        sender = DECLARE_SENDER_ADDRESS
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/deployer/deployer.py:106]: We can add these attributes when are supported. I don't see a point of adding flags that must be equal to a certain value.

File: protostar/commands/declare/declare_command.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                type="str",
+                is_array=True,
+            ),
+            Command.Argument(
+                name="token",
+                description="Used for declaring contracts in Alpha MainNet.",
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/declare/declare_command.py:48]: I copied it from the Starknet CLI's argument description.

File: website/docs/tutorials/guides/testing.md (hunk 2 above)
Comment (kasperski95) [reply website/docs/tutorials/guides/testing.md:2]: Fine.
Personally, I find time hints convenient. I copied this approach from here:
https://developers.google.com/tech-writing/one

File: protostar/deployer/deployer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/deployer/deployer.py:106]: Your call

File: protostar/commands/deploy/deploy_command.py [hunk 8]
```diff
@@ -101,85 +95,28 @@ def arguments(self) -> List[Command.Argument]:
...
     async def run(self, args):
-        return await self.deploy(
+        try:
+            network_config = self._deployer.build_network_config(
+                network=args.network, gateway_url=args.gateway_url
+            )
+        except InvalidNetworkConfigurationException as err:
+            raise ProtostarException(
+                f"Argument `{DeployCommand.gateway_url_arg.name}` or `{DeployCommand.network_arg.name}` is required"
+            ) from err
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/deploy/deploy_command.py:105]: I did it on purpose to indicate the error message should be tailored for each command.

File: protostar/commands/declare/declare_command.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+        return await self.declare(
+            compiled_contract_path=args.contract,
+            network=args.network,
+            gateway_url=args.gateway_url,
+            token=args.token,
+            signature=args.signature,
+        )
+
+    # pylint: disable=too-many-arguments
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/declare/declare_command.py:70]: In a different PR we can increase the threshold, and remove all pylint comments.

File: website/docs/tutorials/guides/testing.md (hunk 2 above)
Comment (mkaput) [reply website/docs/tutorials/guides/testing.md:2]: I am unopposed either to have these hints or not 🤷🏻

File: protostar/commands/declare/declare_command.py (hunk 3 above)
Comment (kasperski95) [reply protostar/commands/declare/declare_command.py:35]: NestJS doesn't use articles e.g.
"Path to webpack configuration."
"The name of the generated component."

https://docs.nestjs.com/cli/usages

---

> If it is required to be concise, e.g. in [headlines](https://en.wikipedia.org/wiki/Headline), signs, labels, and notes, articles are often omitted along with certain other [function words](https://en.wikipedia.org/wiki/Function_word)
https://en.wikipedia.org/wiki/English_articles

File: protostar/commands/declare/declare_command.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/declare/declare_command.py:56]: declare in mainnet, deploy on mainet

File: website/docs/tutorials/guides/deploying.md [hunk 10]
```diff
@@ -4,9 +4,40 @@ sidebar_label: Deploying (1 min)
...
+```sh
+$ protostar declare ./build/main.json --network alpha-goerli
+```
+
+```sh title="The result of running 'protostar declare'."
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply website/docs/tutorials/guides/deploying.md]: Nope. Removed.

File: protostar/starknet_gateway/network_config.py [hunk 11]
```diff
@@ -16,6 +20,24 @@ def __init__(self):
 
 
 class NetworkConfig:
+    class Builder:
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/starknet_gateway/network_config.py:23]: I can't pass `NetworkConfig` as a dependency (unless I pass it as a class, which introduces different issues).

Actually this is simple class. I can use it directly. I'll change it.

**Review by MaksymilianDemitraszek — APPROVED** [review]

File: protostar/commands/declare/declare_command.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    def name(self) -> str:
+        return "declare"
+
+    @property
+    def description(self) -> str:
+        return "Sends a declare transaction to StarkNet."
+
+    @property
+    def example(self) -> Optional[str]:
+        return None
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/declare/declare_command.py:34]: It's easy to forget to update examples, and they are visible only on the website. I don't think they are currently super useful.

**Pull request merged**